### ISSUE TITLE
Any idea to make this cross-platform? (windows more specifically?)

### ISSUE THREAD

**Issue by ggonmar** [will-stone/browserosaurus#368]
I'm browsing for this exact program you've developed but I see it's only for Mac (correct me if I'm wrong)...

Would there be a chance that this becomes a cross-platform thing? (probably not an Issue, but more of a enhancement request...!)

**Comment by will-stone** [comment]
Hi. Browserosaurus will not be cross-platform as I do not use anything other than a Mac. That is unless I buy a non-Apple machine. I suggest forking Browserosaurus. It shouldn’t be too much effort to get it running on Windows, it all depends on if it is possible to scan for installed browsers. I only ask, if you decide to publish, that you rename the project. Thanks 😊

**Comment by ggonmar** [comment]
Will give it a try, thanks!

**Comment by will-stone** [comment]
Nice. Well, let me know how you get on (my contact details are on [wstone.io](https://wstone.io)) and I'm happy to advise where possible. For starters, you essentially just need to swap out the `appExists` package for something that can detect apps on Windows: https://github.com/will-stone/browserosaurus/blob/master/src/utils/filterAppsByInstalled.ts#L10 

Or... enquire with Sindre as to whether he'd be open to an update to his package to support Windows: https://github.com/sindresorhus/app-exists

**Comment by will-stone** [comment]
Hi @rayat-amperity Thank you for considering using Browserosaurus as your base for a windows version. Someone once suggested they’d be open to making a Linux version but they never did, as far as I know. The name I suggested to them was Browserotops but any dinosaur rif would be a nice nod back to this project. What ever name you go for, I’d rather it is tweaked slightly just so that people searching for support do not come here seeking help. Also note that the current license is GPL, so please adhere to all that entails. Good luck, and ping me if you’d like to chat about anything or would like to suggest improvements that both our projects would benefit from. My contact details are on wstone.uk if you’d like to directly contact me.

**Comment by riotrah** [comment]
You have a penchant for puns my friend. 
Gonna park browseratops for now then! 

No clue on what kind of pace at which I'll realistically be able to work on it, so I don't want to claim the monopoly on the name if I end up doing nothing with it reasonably soon. If other community members start making more substantial progress on a fork for another platform, then please feel free to ping me here so I can give up the name!

Thanks for your incredible work on this project, and your gracious support of community efforts like mine <3

**Comment by riotrah** [comment]
Got it "working" on Windows (meaning, I hardcoded the url, but it does indeed run and open my installed browsers in the corresponding private modes)! It _did_ require some admittedly _curious_ decision making on my part, particularly as it pertained to how `apps.ts` makes use of a large map/dict/obj of mac bundle ids -> metadata, but for the few who are interested, it'll prob be obvious from source.

![image](https://user-images.githubusercontent.com/22646419/156683476-2015a0db-d1a0-4125-9dab-fc3713ebc45b.png)

The changes required were a lot more involved than the mere changing of the `appExists` - interestingly this was the easier part - but nothing substantial really.

> I'm not sure you're even interested in this much detail tbh.

However, along the way I did come to the realization that making browseratops a cross platform variant continually built on-top of your project might be the best (read: easiest) way forward for it. I'm not sure if you have any objections to that, but feel free to lmk!

**Comment by will-stone** [comment]
@riotrah Looks great! When you say

> continually built on-top of your project

Do you mean a fork that's rebased onto the upstream project (Browserosaurus)? If so, sure. That's what I thought you'd do. Beware though, when the coding bug gets to me, I do love a refactor. Having said that, it's pretty stable at the moment.

**Comment by riotrah** [comment]
Haha no worries, look forward to learning from whatever overhauls you do. Has been interesting seeing the ones you've already done so far in the git history ;).

Couple questions for when you have time:

1. I'm new to github cd/ci and I wanted to ask - how do you publish releases? Do you do so manually after successful builds? I notice there's no such step in your build workflow.

2. In a nutshell, what does GPL entail? I can also just tie my own shoes and learn about it lol

3. As I was going about working on this, I realized that some changes I make might benefit yours, but only if there's a philosophical (I guess) alignment or whatever:

   - Would you be interested in accepting backports (upstream merges? idk the terminology) that - without affecting Mac compat or introducing any mac regressions - adds multi-platform capabilities (or at least assumes that multi platform is a consideration). Of course, you  would/could make it **_painfully obvious_** that you yourself will not be explicitly supporting or testing anything other than (your version of) mac, and anything else is subject to community or fork support? This is for if in the future you ever change os personally (hey it could happen), but also so that downstream project(s) (like mine) could more easily stay up to date.
   - A concrete example would be something like:
     - Before multi platform assumption: 
       1. you might add some prop to, say, the BrowserWindow constructor, eg `transparent: true`. 
       2. you push, my upstream pull action runs, I merge in yours. 
       3. Then I work around it for windows or change it to whatever windows equivalent exists, or opt-out of whatever you just added etc.
     - After multi platform assumption:
       1. adding the same prop as before now looks like: `transparent: ifMac(true)`
       2. on my end: same pulling and merging as before. 
       4. Then, either I make it `transparent: ifOS(['win', 'mac'], true)` or keep it as-is (`ifMac(true)`), depending on whether your change makes sense for windows or nix etc etc.

This would be added overhead for you for sure, so I don't expect you to be too into it, nor should you be really, given how much backwards bending over you've done so far for the community. Still, thought I'd ask or plant the seed.

Of course, this would be contingent on (aside from your willing participation to start with):

a. Thorough testing added, if not already sufficient, for regression catching 🤞🏽 
b. Enough sleek AF utils and abstractions for the "multi platform assumption" to be as easy and thoughtless as possible for you to make mac-targetted changes🤞🏽 

Finally, unrelated to the multi platform thing, if I ever get into adding other features like URL-based default browsers, or Chrome/ium profiles, would you be interested in backporting, or would that invite having to support shit you don't care about 😅
